### PR TITLE
armin: distance-decay attention bias for local wall-shear (tau_y/z)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The Transolver backbone uses soft-assignment slice attention: N=131,072 mesh points are softly assigned to K=128 "slice tokens" via a learned weighting `W[B, N, K]`. The Q/K attention then operates at slice level `[B, H, K, head_dim]`. This architecture inherently globalises — every point influences every slice, meaning local spatial relationships are only implicitly captured.

Wall shear (`tau_y`, `tau_z`) is dominated by **local velocity gradients at the boundary layer**: the stress at any surface point is strongly determined by nearby surface-normal flow structure. The tau_y/z error is currently 3.7–4× above the AB-UPT target while surface pressure and volume pressure are approaching target. This strongly suggests the model lacks mechanisms to capture local structure.

**Hypothesis**: Adding a learned **distance-decay bias to the attention logits** at the slice-assignment stage will encourage geographically proximate points to share slice tokens, preserving local spatial context. Specifically, after computing the raw attention score `Q @ K^T / sqrt(d)`, add a bias term `b(r_ij) = -α · log(1 + d_ij / r_0)` (logarithmic distance decay, differentiable, bounded) where `d_ij` is the Euclidean distance between point `i`'s 3D coordinate and slice centroid `j`'s mean coordinate, `α` is a learnable scalar, and `r_0` is a fixed scale parameter (set to 0.1, roughly boundary-layer thickness in normalised units).

This adds O(N·K) = O(131072·128) ≈ 16M distance evaluations per forward pass — feasible on GPU, and no extra parameters except `α`. The hypothesis is that forcing locally coherent slice assignments helps the model learn the spatial derivatives that drive tau_y/z.

### Change to make in `target/train.py` (or the relevant model file)

Inspect the model code to find where the slice-assignment attention scores are computed (likely in a `TransolverBlock` or similar). The target change is:

**1. Compute per-point per-slice distances:**
```python
# coords: [B, N, 3] — normalised 3D spatial coordinates (already exist in training data)
# slice_weights: [B, H, N, K] — pre-softmax attention scores

# Compute slice centroids from current weights (detach for efficiency)
with torch.no_grad():
    w_norm = slice_weights.softmax(-1)   # [B, H, N, K]
    slice_centroids = torch.einsum('bhnk,bnd->bhkd', w_norm, coords)  # [B, H, K, 3]

# Distances from each point to each centroid
# coords: [B, N, 3] -> [B, 1, N, 1, 3]
# slice_centroids: [B, H, K, 3] -> [B, H, 1, K, 3]
diffs = coords.unsqueeze(1).unsqueeze(3) - slice_centroids.unsqueeze(2)  # [B, H, N, K, 3]
dist = diffs.norm(dim=-1)  # [B, H, N, K]

# Learnable distance bias
log_dist_bias = -self.alpha * torch.log1p(dist / self.r0)  # [B, H, N, K]
slice_weights = slice_weights + log_dist_bias
```

Where `self.alpha = nn.Parameter(torch.tensor(1.0))` and `self.r0 = 0.1` (fixed).

**2. Add a `--distance-bias` CLI flag** to `train.py` (default `False`) so the baseline control can be run cleanly.

**3. Run two arms:**

```bash
# Arm A: distance-bias enabled
torchrun --standalone --nproc_per_node=8 train.py \
  --agent armin \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --distance-bias \
  --wandb-group armin-local-attn-dist-bias

# Arm B: baseline control (no --distance-bias)
torchrun --standalone --nproc_per_node=8 train.py \
  --agent armin \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --wandb-group armin-local-attn-dist-bias
```

Run both arms in parallel if you have two available 8-GPU nodes, or sequentially if you have one.

**NOTE:** If `target/train.py` on `yi` does not yet have DDP (check for `init_process_group`), cherry-pick commits `bfbe975` and `1a8f7b7` from alphonse's branch `origin/alphonse/depth-scaling-6l-512d` first, then apply your changes. Emma's PR #355 is fixing this officially, but you should apply it to your branch independently.

### Reporting

Report epoch-by-epoch `val_primary/abupt_axis_mean_rel_l2_pct` and all sub-metrics for both arms. Pay special attention to `val_primary/wall_shear_y_rel_l2_pct` and `val_primary/wall_shear_z_rel_l2_pct` — if the hypothesis is working, those should show the largest relative improvement vs Arm B.

Also report the learned value of `alpha` at the end of training (e.g. `model.alpha.item()` at best epoch). If alpha converges near zero, the model is suppressing the bias — note this.

## Baseline (PR #222, W&B run `ut1qmc3i`)

| Metric | Best (epoch 9) |
|---|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **9.2910%** |
| `val_primary/surface_pressure_rel_l2_pct` | 5.8707% |
| `val_primary/wall_shear_rel_l2_pct` | 10.3423% |
| `val_primary/volume_pressure_rel_l2_pct` | 5.8789% |

**Merge bar: 9.291% (lower is better).**

AB-UPT target gaps (the key driver for this hypothesis):
- `wall_shear_y_rel_l2_pct`: estimated ~13–15% vs AB-UPT target 3.65% (~3.7–4×)
- `wall_shear_z_rel_l2_pct`: estimated ~15% vs AB-UPT target 3.63% (~4×)

Reproduce baseline:
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --model-slices 128 --ema-decay 0.999 --lr-warmup-epochs 1
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #356 because Closing — armin is not a yi project student (no senpai-yi-armin pod). The hypothesis (distance-decay attention bias for tau_y/z) is valid and will be reassigned to an actual yi student in a future round..
